### PR TITLE
Adaptive preconditioning

### PR DESCRIPTION
**Pull request by antoine-levitt**
Fix https://github.com/JuliaMolSim/DFTK.jl/issues/31

### CODE REVIEW

**Comment by antoine-levitt** [comment]
So `Optim.jl` deals with it differently, by introducing a method `precondprep!` and doing `precondprep!(P, x); ldiv!(p, P, r)` with the fallback `precondprep!(P, x) = P`. Interfaces are hard.

**Comment by mfherbst** [comment]
I like the `Optim.jl` interface better, since (a) it integrates better with standard Julia: `ldiv!` is really what should be used for this and makes our preconditioners compatible with e.g. `IterativeSolvers.jl`. (b) It is easier to play with preconditioners: One could just use a bare matrix and it worked.

File: src/eigen/diag.jl [hunk 1]
```diff
@@ -50,7 +50,7 @@ function diagonalise_all_kblocks(eigensolver, ham::Hamiltonian, nev_per_kpoint::
         @assert size(guessk) == (length(kpoints[ik].basis), nev_per_kpoint)
 
         results[ik] = eigensolver(kblock(ham, kpt), guessk;
-                                  prec=kblock(prec, kpt), tol=tol,
+                                  prec=prec(ham, kpt), tol=tol,
```
Comment (mfherbst) [inline src/eigen/diag.jl]: Puh ... that took me a second to understand. Could you make a comment to make sure that before `prec` is a type and after an instance.

File: src/eigen/lobpcg_hyper_impl.jl [hunk 2]
```diff
@@ -258,7 +258,7 @@ function LOBPCG(A, X, B=I, precon=I, tol=1e-10, maxiter=100; ortho_tol=2eps(real
             resids[i+nlocked,niter] = norm(R[:,i])
         end
         vprintln(niter, "   ", resids[:, niter])
-        R = precon\R
+        precon(R, new_X, R)
```
Comment (mfherbst) [inline src/eigen/lobpcg_hyper_impl.jl]: See my suggestion to use `ldiv!` as the interface.

File: test/lobpcg.jl [hunk 3]
```diff
@@ -61,7 +61,7 @@ end
     basis = PlaneWaveBasis(model, Ecut, silicon.kcoords, silicon.ksymops; fft_size=fft_size)
     ham = Hamiltonian(basis)
 
-    res = diagonalise_all_kblocks(lobpcg_hyper, ham, 5, tol=1e-8, prec=PreconditionerKinetic(ham, α=0.1),
+    res = diagonalise_all_kblocks(lobpcg_hyper, ham, 5, tol=1e-8,
                                   interpolate_kpoints=false)
```
Comment (mfherbst) [inline test/lobpcg.jl]: Drop linebeak here.

**Comment by antoine-levitt** [comment]
Yeah you're right, and we'll have to do it like that for direct min anyway. Sucks a bit to have to split the computation in two but I'll have to deal with it. I'll do it when I get in front of a keyboard.

**Comment by antoine-levitt** [comment]
How annoying, julia 1.1 doesn't have `ldiv!(I, x)`. Is it OK to require 1.2, or should we hack it?

**Comment by mfherbst** [comment]
Agreed and done. Also I removed PreconditionerNone, because there really is no point in having it. I also made an `AbstractPreconditioner` to have a generic `ldiv!` fallback for the 2-argument version.  Actually now we could also completely drop the 2-argument version, I find it kind of unintuitive anyway.

One issue I still have is that now our LOBPCG implementation is no longer independent of the rest of our code (because it assumes `precondprep!`. If we keep going down that route it might get harder and harder to contribute it back to general package eventually. Just a remark.

**Comment by antoine-levitt** [comment]
Problem with PreconditionerNone is that otherwise we need to pass (ham, kpt) - > I for the prec_type

**Comment by antoine-levitt** [comment]
Not sure about the fallback. Why do we need that? It's the responsability of the individual types to support the two forms, and I even have a PR to base (I'm on my phone and can't find the number now, but it's a recent PR I made) to define the reverse fallback because Nlsolve and DiffEq did it

**Comment by antoine-levitt** [comment]
Precondprep can go into a generic package, it just needs a default fallback that does nothing, and then we explicitly import and override it on our end. That's what happens with Optim

**Comment by mfherbst** [comment]
> Problem with PreconditionerNone is that otherwise we need to pass (ham, kpt) - > I for the prec_type

Ok ... good point ... I use nothing now to explicitly flag that.

**Pull request merged**